File: detach-mpi/test-mpi-detach-loop.c
#include "mpi-detach.h"
#include <mpi.h>
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void sendrecv_detach(const void *sendbuf, int sendcount, MPI_Datatype sendtype,
                     int dest, int sendtag, void *recvbuf, int recvcount,
                     MPI_Datatype recvtype, int source, int recvtag,
                     MPI_Comm comm, MPIX_Detach_callback *callback,
                     void *cbdata) {
  MPI_Request reqs[2];
  MPI_Irecv(recvbuf, recvcount, recvtype, source, recvtag, comm, reqs + 1);
  MPI_Isend(sendbuf, sendcount, sendtype, dest, sendtag, comm, reqs);
  MPIX_Detach_all(2, reqs, callback, cbdata);
//  MPIX_Detach(reqs+1, callback, cbdata);
/*  MPI_Sendrecv(sendbuf, sendcount, sendtype, dest, sendtag, recvbuf, recvcount, recvtype, source, recvtag, comm, MPI_STATUS_IGNORE);
  callback(cbdata);*/
}

int main(int argc, char **argv) {
  int provided;
  MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
  if (provided != MPI_THREAD_MULTIPLE) {
    printf("This code needs MPI_THREAD_MULTIPLE(%i), threadlevel %i was "
           "provided\n",
           MPI_THREAD_MULTIPLE, provided);
    MPI_Finalize();
    return -1;
  }
  int rank, size;
  MPI_Comm comm[5];
  for (int i = 0; i < 5; i++) 
    MPI_Comm_dup(MPI_COMM_WORLD, comm+i);
  
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  int A[] = {1, 2, 3, 4, 5};
  int B[5] = {-1,-1,-1,-1,-1};
  omp_event_handle_t oevent;

#pragma omp parallel num_threads(8)
#pragma omp single
  {
    for (int i = 0; i < 1; i++) {
      int j = i % 5;
#pragma omp task depend(inout : B[j]) detach(oevent)
      {
        omp_event_handle_t ievent;
#pragma omp task depend(inout : B[j]) detach(ievent)
        sendrecv_detach(A + j, 1, MPI_INT, size - rank - 1, 1, B + j, 1,
                        MPI_INT, size - rank - 1, 1, comm[j],
                        (MPIX_Detach_callback *)omp_fulfill_event,
                        (void *)ievent);
#pragma omp task depend(inout : B[j]) detach(ievent)
        {
          printf("task 2 %i: A=%i, B=%i\n", j, A[j], B[j]);
          A[j] += B[j];
          sendrecv_detach(A + j, 1, MPI_INT, size - rank - 1, 2, B + j, 1,
                          MPI_INT, size - rank - 1, 2, comm[j],
                          (MPIX_Detach_callback *)omp_fulfill_event,
                          (void *)ievent);
        }
#pragma omp task depend(inout : B[j]) detach(ievent)
        {
          printf("task 3 %i: A=%i, B=%i\n", j, A[j], B[j]);
          A[j] += B[j];
          sendrecv_detach(A + j, 1, MPI_INT, size - rank - 1, 4, B + j, 1,
                          MPI_INT, size - rank - 1, 4, comm[j],
                          (MPIX_Detach_callback *)omp_fulfill_event,
                          (void *)ievent);
        }
#pragma omp task depend(inout : B[j]) firstprivate(oevent)
        {
          printf("task 4 %i: A=%i, B=%i\n", j, A[j], B[j]);
          A[j] += B[j];
          sendrecv_detach(A + j, 1, MPI_INT, size - rank - 1, 5, B + j, 1,
                          MPI_INT, size - rank - 1, 5, comm[j],
                          (MPIX_Detach_callback *)omp_fulfill_event,
                          (void *)oevent);
        }
      }
    }
#pragma omp taskwait
  }
  printf("%i, %i, %i\n", rank, B[0], B[3]);
  MPI_Finalize();
}

File: detach-mpi/test-mpi-detach-loop.f90
module mpi_detach
    interface
      subroutine mpix_detach (request, cb, event, ierr)
        use omp_lib_kinds
        integer :: request, ierr
        external :: cb
        integer (kind=omp_event_handle_kind), &
          value, intent(in) :: event
      end subroutine mpix_detach
      subroutine mpix_detach_all (count, requests, cb, event, ierr)
        use omp_lib_kinds
        integer :: requests(*)
        integer :: count, ierr
        external :: cb
        integer (kind=omp_event_handle_kind), &
          value, intent(in) :: event
      end subroutine mpix_detach_all
      subroutine mpix_detach_task (request, event, ierr)
        use omp_lib_kinds
        integer :: request, ierr
        integer (kind=omp_event_handle_kind), &
          value, intent(in) :: event
      end subroutine mpix_detach_task
      subroutine mpix_detach_all_task (count, requests, event, ierr)
        use omp_lib_kinds
        integer :: requests(*)
        integer :: count, ierr
        integer (kind=omp_event_handle_kind), &
          value, intent(in) :: event
      end subroutine mpix_detach_all_task
    end interface
end module
program test

    use omp_lib
    !use mpi_detach

    implicit none

    include "mpif.h"

    integer ::base=0, tmp_int=0, rbuf, i, j
    integer(omp_event_handle_kind) :: oevent, ievent
    integer :: request, comm(0:4)

    integer ::ierr, irank, isize, provided

    integer(kind = mpi_address_kind)::wsize, targ_disp
    integer :: A(0:4)
    integer :: B(0:4)
    A = (/1,2,3,4,5/)
    B = -1

    CALL MPI_INIT_THREAD(MPI_THREAD_MULTIPLE, provided, ierr)
    IF (provided .LT. MPI_THREAD_MULTIPLE) THEN
        WRITE(*,'(A)') 'The threading support level is lesser than that demanded.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierr)
    END IF


    CALL MPI_COMM_RANK(MPI_COMM_WORLD, irank, ierr)

    CALL MPI_COMM_SIZE(MPI_COMM_WORLD, isize, ierr)
    DO i=0,4
      CALL MPI_COMM_DUP(MPI_COMM_WORLD, comm(i), ierr)
    ENDDO

    print *, "Hello from pe", irank, base

    !$omp parallel num_threads(3) private(ierr)
    !$omp single
    DO i=0,14
      j = modulo(i,5)
    !$omp task detach(oevent) depend(inout:A(j)) firstprivate(j, ievent)
          print *, "Outer Task j=", j
		!$omp task detach(ievent) depend(inout:A(j))
		  print *, "Inner Task1 j=", j
          call sendrecv_detach(A(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               B(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               comm(j), omp_fulfill_event, ievent, ierr)
		!$omp end task
		!$omp task detach(ievent) depend(inout:A(j))
		  print *, "Inner Task2 j=", j
		  A(j) = A(j) + B(j)
          call sendrecv_detach(A(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               B(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               comm(j), omp_fulfill_event, ievent, ierr)
		!$omp end task
		!$omp task detach(ievent) depend(inout:A(j))
		  print *, "Inner Task3 j=", j
		  A(j) = A(j) + B(j)
          call sendrecv_detach(A(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               B(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               comm(j), omp_fulfill_event, ievent, ierr)
		!$omp end task
		!$omp task firstprivate(oevent) depend(inout:A(j))
		  print *, "Inner Task4 j=", j
		  A(j) = A(j) + B(j)
          call sendrecv_detach(A(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               B(j), 1, MPI_INTEGER, isize-irank-1, 1, &
                               comm(j), omp_fulfill_event, oevent, ierr)
		!$omp end task

    !$omp end task
    ENDDO
    !$omp taskwait
    !$omp end single
    !$omp end parallel
    
    print *, A(0), A(1), A(2), A(3), A(4), B    

    print *, "Goodbye from pe", irank, base

    CALL MPI_FINALIZE(ierr)

end program

subroutine sendrecv_detach(sendbuf, sendcount, sendtype, dest, sendtag, &
                           recvbuf, recvcount, recvtype, source, recvtag, &
                           comm, callback, cbdata, ierr)
    use omp_lib_kinds
    use mpi_detach
implicit none
INTEGER :: sendbuf(*), recvbuf(*)
INTEGER :: sendcount, sendtype, dest, sendtag
INTEGER :: recvcount, recvtype, source, recvtag, comm
INTEGER :: ierr
external :: callback
INTEGER (kind=omp_event_handle_kind), intent(in) :: cbdata
INTEGER :: reqs(0:1)
call MPI_Irecv(recvbuf, recvcount, recvtype, source, recvtag, comm, reqs(1), ierr);
call MPI_Isend(sendbuf, sendcount, sendtype, dest, sendtag, comm, reqs(0), ierr);
call MPIX_Detach_all_task(2, reqs, cbdata, ierr);
end subroutine sendrecv_detach

